### PR TITLE
Build Asterisk in same FreePBX container 

### PR DESCRIPTION
**Pull request by Stell0**
https://github.com/nethesis/ns8-nethvoice/issues/101
- Asterisk in same FreePBX container
- Instead of using Asterisk CentOS 7 rpm, now Asterisk is built during container build
- now it's based on Debian
- Added [Supervisord](http://supervisord.org/index.html) to FreePBX container
- Supervisor runs Asterisk and Apache
- FreePBX modules initialization moved to anther script because needs Asterisk to be started before

### CODE REVIEW

File: freepbx/README.md [hunk 1]
```diff
@@ -43,6 +43,18 @@ This container also hosts push-proxy script for Flexisip.
...
+## Asterisk Environment variables
+
+- `ASTMANAGERHOST` is the ip where AMI , Asterisk Manager Interface is exposed. 127.0.0.1 in network=host configuration
+- `ASTMANAGERPORT` port of AMI, Asterisk Manager Interface
+- `AMPMGRUSER` User automatically configured to access ti AMI
+- `AMPMGRPASS` Password of AMI user
+- `NETHVOICE_MARIADB_PORT` Port of MariaDB
+- `ASTERISK_RTPSTART` and `ASTERISK_RTPEND` are the UDP port range for RTP packages
+- `ASTERISK_SIP_PORT` and `ASTERISK_SIPS_PORT` are the UDP and TCP ports for SIP transport
+- `ASTERISK_IAX_PORT is the UDP port for IAX transport
```
Comment (Amygos) [inline freepbx/README.md:55]:
```suggestion
- `ASTERISK_IAX_PORT` is the UDP port for IAX transport
```

File: freepbx/etc/apache2/sites-enabled/000-default.conf [hunk 2]
```diff
@@ -15,32 +15,10 @@
...
-
         ErrorLog ${APACHE_LOG_DIR}/error.log
         CustomLog ${APACHE_LOG_DIR}/access.log combined
-
-        # For most configuration files from conf-available/, which are
-        # enabled or disabled at a global level, it is possible to
-        # include a line for only one particular virtual host. For example the
-        # following line enables the CGI configuration for this host only
-        # after it has been globally disabled with "a2disconf".
-        #Include conf-available/serve-cgi-bin.conf
```
Comment (Amygos) [inline freepbx/etc/apache2/sites-enabled/000-default.conf:44]: Why did you remove the comments?

File: ui/yarn.lock
Comment (Amygos) [inline ui/yarn.lock:1]: This fix is already present in the `main` branch https://github.com/nethesis/ns8-nethvoice/commit/3cd0721bc05120b6d69d3046aba94da49babf719

File: imageroot/systemd/user/freepbx.service [hunk 3]
```diff
@@ -64,6 +69,9 @@ ExecStart=/usr/bin/podman run \
     --hostname=${NETHVOICE_HOST} \
     ${NETHVOICE_FREEPBX_IMAGE}
 
+# Wait till asterisk is active
+ExecStartPost=podman exec asterisk bash -c "while [[ $$(/usr/sbin/asterisk -rx 'core show version' 2>/dev/null) != Asterisk* ]]; do ((++attempt<30)) || exit 2; sleep 1 ; done"
```
Comment (Amygos) [inline imageroot/systemd/user/freepbx.service:73]: Why do you need to wait?

**Review by Amygos — COMMENTED** [review]
Just my 2 cent before the PR is ready

File: imageroot/systemd/user/freepbx.service (hunk 3 above)
Comment (Stell0) [reply imageroot/systemd/user/freepbx.service:73]: to make sure asterisk is full started before running dependent services

File: freepbx/etc/apache2/sites-enabled/000-default.conf (hunk 2 above)
Comment (Stell0) [reply freepbx/etc/apache2/sites-enabled/000-default.conf:44]: to improve configuration readability

File: freepbx/Containerfile [hunk 4]
```diff
@@ -1,169 +1,409 @@
...
+# Set fixed Asterisk version
+ARG ASTERISK_VERSION=18.21.0
+ARG PJSIP_VERSION=2.13.1
+ARG DEBIAN_FRONTEND=noninteractive
+ARG BUILDROOT=/usr/src/asterisk/build
+ARG LIBDIR=/usr/lib
+ARG DATADIR=/var/lib
+ARG MAKEARGS="DEBUG= OPTIMIZE= DESTDIR=${BUILDROOT} ASTVARRUNDIR=/var/run/asterisk ASTDATADIR=${DATADIR}/asterisk ASTVARLIBDIR=${DATADIR}/asterisk ASTDBDIR=${DATADIR}/asterisk/db NOISY_BUILD=1"
+COPY usr/src/asterisk /usr/src/asterisk
+RUN apt-get update && \
```
Comment (Amygos) [inline freepbx/Containerfile:13]: Why did you put all the commands into a single `RUN`? In this way, every time you change something the layer will be discarded and regenerated, and the cache will not be used.

File: freepbx/entrypoint.sh [hunk 5]
```diff
@@ -5,6 +5,93 @@
...
+noload = res_stir_shaken.so
+noload = res_pjsip_stir_shaken.so
+noload = res_pjsip_phoneprov.so
+noload = res_pjsip_phoneprov_provider.so
+noload = cdr_csv.so
+noload = cdr_syslog.so
+noload = app_alarmreceiver.so
+noload = res_http_media_cache.so
+noload = res_phoneprov.so
+EOF
```
[the remark this answers is not in the document]
Comment (Stell0) [reply freepbx/entrypoint.sh:75]: overwriting `asterisk.conf` you can update module from the previous version
`modules.conf` is overwritten by FreePBX if an Asterisk module fails to load, and it wont load ever again. This make sure that restarting container modules.conf isn't messed up by FreePBX

File: freepbx/Containerfile (hunk 4 above)
Comment (Stell0) [reply freepbx/Containerfile:13]: To reduce the number of the layers.
When this is in production, I expect that we will change this only on Asterisk version update.

Comment (Stell0) [reply freepbx/Containerfile:13]: > Ok, but this is a multistage container and is not the builder container that will be shipped.

I've added a RUN after the apt-get

**Review by Amygos — APPROVED** [review]
